### PR TITLE
Autobuild: Create Debian repository on release

### PR DESCRIPTION
**Pull request by ann0see**
**Short description of changes**

* This PR creates files in the autobuild process to allow apt to use a GitHub release as apt repository. Based on work by @hoffie from https://github.com/jamulussoftware/repos and this idea: https://github.com/jamulussoftware/repos/issues/2#issuecomment-1428795106
* Tested and working based on this release: https://github.com/ann0see/jamulus/releases/tag/r3_9_1_repotest2
* For latest release test see: https://github.com/ann0see/jamulus/releases/tag/r3_9_1devrepo (tested and working)

I consider this approach easier as it is more automated and there's no need for a separate repo.

CHANGELOG: Linux: Debian users can now use the official PPA to get automatic updates. See the Linux install page on jamulus.io for more information.

**Context: Fixes an issue?**

Related to: https://github.com/jamulussoftware/jamulus/discussions/2122

**Does this change need documentation? What needs to be documented and how?**
Yes. We need to supply the link to the repository, probably using the redirect to the latest release by GitHub and a link to the GPG key on jamulus.io

**Status of this Pull Request**

Ready for review. Before this can be merged, we need to add a GPG key as secret.

**What is missing until this pull request can be merged?**

Add GPG key as secret.

## Checklist

-  [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/main/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
-  [x] I tested my code and it does what I want (pushed a full release after opening this PR, tested headless build)
-  [x] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/main/CONTRIBUTING.md#source-code-consistency) 
-  [x] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
-  [x] I've filled all the content above

### CODE REVIEW

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml:463]: Forks which don't have a GPG key set, will have a failed build. 
I'm not sure if a silent failure would be better. Skipping the repo build should be possible.

There are multiple options:
* Remove the whole job (easy)
* Introduce a variable enabling/disabling the repo creation
* If GPG_PRIVATE_KEY is not present, skip the job. We still need to spin up the container, I'm afraid. GitHub dosn't support checking for present secrets in the .yaml file directly as far as I understand

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml:470]: While this approach is not perfect, I think it's easier to download all artifacts instead of one step per deb file. In future wildcards might be supported.

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml:477]: Could be in a separate .sh file. Hoffie once said he's in favor of inlining.

File: linux/setup_repo.sh
Comment (ann0see) [inline linux/setup_repo.sh:15]: This uses a trick: latest points to the last release tagged as "stable" by GitHub. Thus, the link remains unchanged, but a redirect guarantees the correct source.

File: linux/setup_repo.sh
Comment (ann0see) [inline linux/setup_repo.sh:1]: Where should we host this file? Is this repo good or should it be on the website?

File: .github/workflows/autobuild.yml
Comment (pljones) [reply .github/workflows/autobuild.yml:463]: Will `needs.create_release.outputs.publish_to_release == 'true'` be true?  Is that what you mean by "failed build"?

File: .github/workflows/autobuild.yml
Comment (pljones) [reply .github/workflows/autobuild.yml:477]: It's easier to review in context although it makes the `autobuild.yml` grow as it becomes more feature-rich.  Maybe we need to have a think about it.

File: .github/workflows/autobuild.yml
Comment (ann0see) [reply .github/workflows/autobuild.yml:463]: Yes. If they publish a release it will be true.

Comment (ann0see) [reply .github/workflows/autobuild.yml:463]: Ok. Maybe my new approach works.

**Comment by ann0see** [comment]
@pljones Do you know anyone who knows a bit more about Debian packaging or someone who can review this approach? 
Hoffie is quite experienced with bash scripting, and the code looks ok. Nevertheless we should have a few eyes looking over this.

**Comment by pljones** [comment]
Sorry, no - for Linux, I've pretty much been a consumer for packaging technologies.  I've no development experience there, nor know anyone who has.  (At work we use RPMs and don't even use standard deployment for those, so totally unhelpful here!)

**Comment by ann0see** [comment]
Ok. mirabilos was the Debian maintainer, but he stepped back so I'm not sure if he'd want to review this. ignotus has some knowledge of bash scripts, but I don't think he knows deb packaging. @trebmuh might know a bit as he develops a distribution.

**Comment by softins** [comment]
I've seen this review request, and will review it in the next day or two.

**Comment by ann0see** [comment]
Thank you!

File: .github/workflows/autobuild.yml
Comment (softins) [reply .github/workflows/autobuild.yml:463]: That run demonstrates that skipping the job works if the GPG_PRIVATE_KEY is not set. I can't see a run that demonstrates the action is working properly when the private key IS set.

I'm assuming that the main jamulus repo has a private key, but the ann0see repo does not. But to test this before merging, I guess either ann0see's repo needs a (temporary?) private key, or perhaps we need to merge these changes into a branch on the jamulus repo, so that it can use the jamulus repo secrets?

I don't profess to be an expert (or even moderately knowledgeable) in Debian packaging or repos. Most of my packaging experience is with RPMs on RH-based systems.

File: linux/setup_repo.sh
Comment (softins) [inline linux/setup_repo.sh:17]: Just trying this script locally, and this step fails with a 404 for `key.asc`.

Is this script intended for use by a Github workflow, or by an end user? I can't see it referred to anywhere else in the tree.

Comment (pljones) [reply linux/setup_repo.sh:17]: > Is this script intended for use by a Github workflow, or by an end user? I can't see it referred to anywhere else in the tree.

Purely by our git build to create a repo for apt to auto update from.

Comment (ann0see) [reply linux/setup_repo.sh:17]: You'd need to replace the variables. The release will generate a runnable script.

**Comment by ann0see** [comment]
Also my repo has a GPG key set.

File: .github/workflows/autobuild.yml
Comment (ann0see) [reply .github/workflows/autobuild.yml:463]: See https://github.com/ann0see/jamulus/releases/tag/r3_9_1devrepo
That's a working (?) repo

File: linux/setup_repo.sh
Comment (softins) [reply linux/setup_repo.sh:17]: I've currently checked out `ann0see/debrepo`, but I can't see anywhere that calls `setup_repo.sh`. I would have expected it to be referred to somewhere within `autobuild.yml`? Or within another script somewhere in the tree?

You can tell I'm a bit rusty!

Comment (ann0see) [reply linux/setup_repo.sh:17]: setup_repo.sh is a script which the user should run on his machine to install the repository. The logic to create the repository is in autobuild.yml. I believe the content in autobuild.yaml correctly replaces the content of the setup_repo.sh script and uploads the modified version to the release.

**Review by pljones — APPROVED** [review]

**Comment by pljones** [comment]
Just checking: is there anything outstanding on this?  Rebase?

Does it push beta and pre releases to the repo?  (I forgot to check.)

**Comment by ann0see** [comment]
The standard URL will point to the latest stable release. So even though the files for the Debian repo will get uploaded to beta releases users would need to set the respective direct link to beta/nightly releases.

**Comment by ann0see** [comment]
@softins Are you OK with a merge? Documentation will be added.

**Review by softins — APPROVED** [review]
> @softins Are you OK with a merge? Documentation will be added.

Well it looks sensible, but I'm not able to do any testing myself at the moment. Since @pljones was happy to approve, I'll add my approval too.

**Comment by ann0see** [comment]
Ok. Will do a rebase and then merge it. We can always state that this is still experimental

**Comment by ann0see** [comment]
Dropping needs documentation as I've opened https://github.com/jamulussoftware/jamuluswebsite/issues/937 on the website repository.

**Pull request merged**
